### PR TITLE
Fix gather proxy extreme MPI issue

### LINKED ISSUE #21968: Diffing tests on 9 ranks
## Bug Description
Two tests fail only when using 9 MPI ranks (while passing with any other number of ranks up to 16).

```
[0.530s]     DIFF postprocessors/element_extreme_value.proxy_elemental_extreme FAILED (CSVDIFF)
[0.510s]     DIFF postprocessors/nodal_extreme_value.nodal_extreme_proxy FAILED (CSVDIFF)
```

## Steps to Reproduce
Enable the (extremely long) "Parallel sweep" test. Or check https://civet.inl.gov/job/1157523/

## Impact
Eliminate a potential rare MPI bug.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #21968

The SFINAE stuff checks if a `<` operator exists for the proxy type. If not, use the old way (sorry, no good options here). If it does then we need to sort on the pair (value,proxy). Unfortunately `maxloc`/`minloc` do not work for non POD types :-(. So we just allgather the pairs and find the max/min element ourselves. On the upside it is just one MPI call :-D

### CODE REVIEW

**Comment by dschwen** [comment]
We might need to use a similar comparison strategy for multi-threaded execution...

**Comment by dschwen** [comment]
LOL, this builds only on clang?!

**Review by loganharbour — COMMENTED** [review]
Just use `if constexpr` with one definition as to avoid the SFINAE madness?

**Comment by dschwen** [comment]
I'll try but I'm not sure it is possible in this case, as the < operator would give an error for some types

**Comment by dschwen** [comment]
By the way the errors I'm getting on gcc are

```
/home/schwd/Programs/moose-9rank-fail/framework/build/header_symlinks/UserObject.h:260:1: error: no declaration matches 'typename std::enable_if<std::is_same<decltype ((proxy < proxy)), bool>::value, void>::type UserObject::gatherProxyValueMax(T1&, T2&)'
  260 | UserObject::gatherProxyValueMax(T1 & value, T2 & proxy) ->
      | ^~~~~~~~~~
/home/schwd/Programs/moose-9rank-fail/framework/build/header_symlinks/UserObject.h:165:8: note: candidates are: 'template<class T1, class T2> typename std::enable_if<std::is_same<decltype ((proxy < proxy)), bool>::value, void>::type UserObject::gatherProxyValueMax(T1&, T2&)'
  165 |   auto gatherProxyValueMax(T1 & value, T2 & proxy) ->
      |        ^~~~~~~~~~~~~~~~~~~
/home/schwd/Programs/moose-9rank-fail/framework/build/header_symlinks/UserObject.h:162:8: note:                 'template<class T1, class T2> typename std::enable_if<(! std::is_same<decltype ((proxy < proxy)), bool>::value), void>::type UserObject::gatherProxyValueMax(T1&, T2&)'
```

to make this a bit more clear:

```
                     error: no declaration matches 'typename std::enable_if<std::is_same<decltype ((proxy < proxy)), bool>::value, void>::type UserObject::gatherProxyValueMax(T1&, T2&)'
note: candidates are: 'template<class T1, class T2> typename std::enable_if<std::is_same<decltype ((proxy < proxy)), bool>::value, void>::type UserObject::gatherProxyValueMax(T1&, T2&)'
```

the implementation is missing the `template<class T1, class T2>` which is weird, because ... it is clearly there.

**Comment by dschwen** [comment]
> Just use `if constexpr` with one definition as to avoid the SFINAE madness?

Yeah, this actually seems to work.

**Comment by dschwen** [comment]
> > Just use `if constexpr` with one definition as to avoid the SFINAE madness?
> 
> Yeah, this actually seems to work.

But it shouldn't... (see here https://www.codingame.com/playgrounds/2205/7-features-of-c17-that-will-simplify-your-code/constexpr-if where numeric limits for a string are said to give a compile error). I guess we are not using this for non comparable types after all. I must have misread a previous compile error (that was probably just the inability to use maxloc with pair types).

**Comment by dschwen** [comment]
@loganharbour see https://godbolt.org/z/rjs1KGP6W, template substitution fails if the condition in constexpr if doesn't compile.

**Comment by dschwen** [comment]
Fix breaks distributed meshes. :-(

File: framework/include/utils/MooseUtils.h [hunk 1]
```diff
@@ -1065,6 +1065,25 @@ get(const std::shared_ptr<T> & s)
   return s.get();
 }
 
+// print a message from all processors (use only for debugging purposes)
+template <typename... Args>
+void
+parallelPrint(const Parallel::Communicator & comm, const std::string & title, Args &&... args)
```
Comment (dschwen) [inline framework/include/utils/MooseUtils.h:1071]: This came in handy for parallel debugging and beats `--keep-cout --redirect-stdout` plus `grep`, IMO

File: framework/src/postprocessors/ElementExtremeValue.C [hunk 2]
```diff
@@ -53,56 +54,52 @@ ElementExtremeValue::initialize()
...
 ElementExtremeValue::getValue()
 {
-  return _value;
+  return _proxy_value.second;
 }
 
 void
 ElementExtremeValue::finalize()
 {
+  MooseUtils::parallelPrint(_communicator, _name, _proxy_value.first, ", ", _proxy_value.second);
```
Comment (dschwen) [inline framework/src/postprocessors/ElementExtremeValue.C:95]: I'll remove this if the tests should pass now...

File: framework/include/utils/MooseUtils.h (hunk 1 above)
Comment (loganharbour) [reply framework/include/utils/MooseUtils.h:1071]: `SerializerGuard` to cerr does this too :-)

**Review by loganharbour — DISMISSED** [review]
LGTM 👍🏼

File: framework/include/utils/MooseUtils.h (hunk 1 above)
Comment (dschwen) [reply framework/include/utils/MooseUtils.h:1071]: No, it does not.

Comment (dschwen) [reply framework/include/utils/MooseUtils.h:1071]: To clarify, if you use `std::cout` with a barrier you still need to supply `--keep-cout`, which gives you _all_ console output N times.

**Comment by dschwen** [comment]
I think I want to clean this up a bit more though

File: framework/include/utils/MooseUtils.h (hunk 1 above)
Comment (dschwen) [reply framework/include/utils/MooseUtils.h:1071]: oh... good point.

**Comment by dschwen** [comment]
@loganharbour This should be ready for another pass now. I refactored `ElementExtremeValue`, `ElementExtremeFunctorValue`, and `NodalExtremeValue` to use a common base class that does all extreme value and proxy value handling. I also restored functionality to use the simpler `gatherMin`/`gatherMax` if no proxy value is used.

**Comment by dschwen** [comment]
That coverage fail seems unrelated

```
>> Merging the head repo into the base into ref "884e13d498d3b8e6510d5a44fece71801a32549c_covhead_sSlAAf"
remote: Internal Server Error.
remote: 
fatal: unable to access 'https://github.com/dschwen/moose.git/': The requested URL returned error: 500
/data/civet3/build/moose: git checkout -q -b 884e13d498d3b8e6510d5a44fece71801a32549c_covhead_sSlAAf
/data/civet3/build/moose: git pull -q --no-ff https://github.com/dschwen/moose.git 884e13d498d3b8e6510d5a44fece71801a32549c
An error occurred when running the command "git pull -q --no-ff https://github.com/dschwen/moose.git 884e13d498d3b8e6510d5a44fece71801a32549c"
```

**Review by loganharbour — APPROVED** [review]

**Pull request merged**